pkl-core/src/main/java/org/pkl/core/runtime/TestRunner.java

```diff
@@ -387,74 +388,85 @@ private static String getDisplayUri(ModuleInfo moduleInfo) {
   }
 
   private static Failure factFailure(SourceSection sourceSection, String location) {
-    return new Failure("Fact Failure", sourceSection.getCharacters().toString(), location);
+    String message = sourceSection.getCharacters().toString() + " " + renderLocation(location);
+    return new Failure("Fact Failure", message);
   }
 
   private static Failure exampleLengthMismatchFailure(
       String location, String property, int expectedLength, int actualLength) {
     String msg =
-        "Output mismatch: Expected \""
+        renderLocation(location)
+            + "\n"
+            + "Output mismatch: Expected \""
             + property
             + "\" to contain "
             + expectedLength
             + " examples, but found "
             + actualLength;
 
-    return new Failure("Output Mismatch (Length)", msg, location);
+    return new Failure("Output Mismatch (Length)", msg);
   }
 
   private static Failure examplePropertyMismatchFailure(
       String location, String property, boolean isMissingInExpected) {
 
-    String exists_in;
-    String missing_in;
+    String existsIn;
+    String missingIn;
 
     if (isMissingInExpected) {
-      exists_in = "actual";
-      missing_in = "expected";
+      existsIn = "actual";
+      missingIn = "expected";
     } else {
-      exists_in = "expected";
-      missing_in = "actual";
+      existsIn = "expected";
+      missingIn = "actual";
     }
 
-    String err =
-        "Output mismatch: \""
+    String message =
+        renderLocation(location)
+            + "\n"
+            + "Output mismatch: \""
             + property
             + "\" exists in "
-            + exists_in
+            + existsIn
             + " but not in "
-            + missing_in
+            + missingIn
             + " output";
 
-    return new Failure("Output Mismatch", err, location);
+    return new Failure("Output Mismatch", message);
   }
 
   private static Failure exampleFailure(
       String location,
       String expectedLocation,
       String expectedValue,
       String actualLocation,
-      String actualValue) {
+      String actualValue,
+      int exampleNumber) {
     String err =
-        "Expected: "
-            + "("
-            + expectedLocation
-            + ")"
-            + "\n"
-            + expectedValue
-            + "\n"
+        "#"
+            + exampleNumber
+            + " "
+            + renderLocation(location)
+            + ":\n  "
+            + "Expected: "
+            + renderLocation(expectedLocation)
+            + "\n  "
+            + expectedValue.replaceAll("\n", "\n  ")
+            + "\n  "
             + "Actual: "
-            + "("
-            + actualLocation
-            + ")"
```

Comment:
Here I think you need an extra space in all of the `"\n  "`, as mentioned in the example output, so that it aligns with the "`(` column" (i.e. 3 spaces following the new line instead of the current two spaces).

Reply:
The +1 space IMHO is not really "inconsistent" but a side-effect because the "fields" are not fixed-width. So, when you have "multiple fields" in one line (`#<nth>` and the `(file:...)`) and the fields are not fixed-width, then it's inevitable.

That's also why I think the "inconsistent" spacing is OK, because we use the inconsistencies to actually make it nice / aesthetically pleasing by aligning to the natural vertical whitespace.

Re. the width of numbers, you are right, I didn't consider it!  But, even so, since most cases will probably have well under 100 cases you can easily cover 90%+ of the actual usage by accounting for double digits.

E.g.

```
          examples
            ✅ user 0
            ❌ user 1
               #1  (file:///file, line x):
                   Expected: (file:///file, line x)
                     new {
                       name = "Parrot"
                       age = 35
                     }
                   Actual: (file:///file, line x)
                     new {
                       name = "Welma"
                       age = 35
                     }
               #91 (file:///file, line x):
                   Expected: (file:///file, line x)
                     new {
                       name = "Parrot"
                       age = 35
                     }
                   Actual: (file:///file, line x)
                     new {
                       name = "Welma"
                       age = 35
                     }
        ❌ 50.0% tests pass [2/4 failed], 66.7% asserts pass [2/6 failed]

```

That would almost surely work for most cases and I think it looks muuuch better than the `E/A` from `Expected` / `Actual` "mis-aligned` and taking over the vertical whitespace ... that really triggers me 😅

But I agree, a big part of this is a matter of preference, so  🤷.

Reply:
I think I'm going to keep my version as-is--not totally sold that indenting this much is a good thing. Sorry! But yes, there's no right way to do this, and I don't think my version is necessarily better than yours. However, I am concerned about the amount of indentation we already have, and I think we can improve this output even more when we have text coloring.

Reply:
Cool, no worries, thanks for considering it!